tools/single_use/SaveUserIDToInstrumentData.php

```diff
@@ -0,0 +1,147 @@
+<?php
+/**
+  * RUN-TIME WARNING FOR PROJECTS: It can take up to a whole day to run this script. It took 14 hours to run the script on the CCNA DB during testing.
+  *
+  * This script is intended for a one-time use only to restore the value of the
+ * `UserID` column of instrument tables and the `UserID` key of the instrument
+ * JSON `Data` in the flag table.
+ *
+ * This tool queries data from the `history` table and meaningfully imports
+ * data from its `userID` column back into the instrument and `flag` tables.
+ *
+ * Example use:
+ * $ php SaveUserIDToInstrumentData.php [confirm]
+
+ * PHP Version 7
+ *
+ * @category Tools
+ * @package  Loris
+ * @author   Zaliqa Rosli <zaliqa.rosli@mcin.ca>
+ * @licence  LORIS License
+ * @link     https://www.github.com/aces/Loris
+ */
+require_once __DIR__ . '/../generic_includes.php';
+
+$confirm = false;
+if (isset($argv[1]) && $argv[1] === 'confirm') {
+    $confirm = true;
+}
+
+echo "\n\nQuerying data...\n\n";
+
+// Get history DB
+$DBInfo = $config->getSetting('database');
+$hDB = isset($DBInfo['historydb']) ? $DBInfo['historydb'] : $DBInfo['database'];
+
+$result_count = 1;
+
+// Query entries in the history table for the latest change
+// made to the flag table, for every CommentID.
+$history = $DB->pselect(
+    "SELECT f.Test_name, f.CommentID, h1.userID
+      FROM flag f
+      JOIN $hDB.history h1 on (f.CommentID=h1.primaryVals)
+      JOIN
+        (
+          SELECT primaryVals, MAX(changeDate) as changeDate
+            FROM $hDB.history
+            WHERE userID <> 'unknown'
+              AND type='U'
+              AND NOT (
+                col='Data_entry' AND old IS NULL
+              )
+            GROUP BY primaryVals
+        ) h2 USING (primaryVals, changeDate)
+            GROUP BY CommentID, h1.userID",
```

Comment:
This query is fairly complex and I haven't looked at it closely enough yet to make sure it's correct, but the `$hDB` value should be quoted.

Reply:
why does the $hDB value need to be quoted? just wondering because i dont get an error when the script runs the query as is

Reply:
in case there are SQL special characters i nthe DB name. its an issue we have had in the past with other projects

Reply:
okayy, so by quote, do we mean sql backticks or actual quote

Reply:
SQL backticks, there might even be a function for it in the database class... escape() i think